src/perf/lib/PerfHelpers.h

```diff
@@ -0,0 +1,294 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    MsQuic API Perf Helpers
+
+--*/
+
+#pragma once
+
+#ifdef QUIC_CLOG
+#include "PerfHelpers.h.clog.h"
+#endif
+
+#include <msquic.h>
+
+class QuicApiTable;
+
+extern const QuicApiTable* MsQuic;
+
+#define QUIC_SKIP_GLOBAL_CONSTRUCTORS
+
+#include <quic_platform.h>
+#include <msquic.hpp>
+
+#include <msquichelper.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#ifndef _KERNEL_MODE
+#include <new> // Needed for placement new
+#else
+#include <new.h>
+#endif
+
+#ifdef _KERNEL_MODE
+extern uint8_t SelfSignedSecurityHash[20];
+#else
+extern QUIC_SEC_CONFIG_PARAMS* SelfSignedParams;
+#endif
+extern bool IsSelfSignedValid;
+
+#define QUIC_TEST_SESSION_CLOSED    1
+
+inline
+int
+WriteOutput(
+    _In_z_ const char* format
+    ...
+    )
+{
+#ifndef _KERNEL_MODE
+    va_list args;
+    va_start(args, format);
+    int rval = vprintf(format, args);
+    va_end(args);
+    return rval;
+#else
+    UNREFERENCED_PARAMETER(format);
+    return 0;
+#endif
+}
+
+struct MsQuicListener {
+    HQUIC Handle{ nullptr };
+    QUIC_LISTENER_CALLBACK_HANDLER Handler;
+    void* Context;
+    MsQuicListener(const MsQuicSession& Session) {
+        if (!Session.IsValid()) {
+            return;
+        }
+        if (QUIC_FAILED(
+            MsQuic->ListenerOpen(
+                Session,
+                [](HQUIC Handle, void* Context, QUIC_LISTENER_EVENT* Event) -> QUIC_STATUS {
+                    MsQuicListener* Listener = (MsQuicListener*)Context;
+                    return Listener->Handler(Handle, Listener->Context, Event);
+                },
+                this,
+                &Handle))) {
+            Handle = nullptr;
+        }
+    }
+    ~MsQuicListener() noexcept {
+        if (Handler != nullptr) {
+            MsQuic->ListenerStop(Handle);
+        }
+        if (Handle) {
+            MsQuic->ListenerClose(Handle);
+        }
+    }
+
+    QUIC_STATUS
+    Start(
+        _In_ QUIC_ADDR* Address,
+        _In_ QUIC_LISTENER_CALLBACK_HANDLER ShadowHandler,
+        _In_ void* ShadowContext) {
```

Comment:
Instead of Shadow in the name, I prefer either you prefix with _ or you use the same name, and when you set it, do something like `this->Handler = Handler`.

Reply:
Also, this might make more sense to pass these in the constructor anyways.

Reply:
It has to be passed here, since because of how construction works I don't have access to these initially.